test/openssl/test_pkey.rb

```diff
@@ -244,4 +244,210 @@ def test_to_text
     rsa = Fixtures.pkey("rsa1024")
     assert_include rsa.to_text, "publicExponent"
   end
+
+  if openssl?(3, 0, 0)
+    def test_from_parameters_with_invalid_alg
+      e = assert_raise(OpenSSL::PKey::PKeyError) {
+        OpenSSL::PKey.from_parameters("ASR", {})
+      }
+      assert_match(/^EVP_PKEY_CTX_new_from_name: unsupported/, e.message)
+    end
+
+    def test_s_from_parameters_rsa_with_n_e_and_d_given_as_integers
+      new_key = OpenSSL::PKey.from_parameters("RSA", n: 3161751493,
+                                                     e: 65537,
+                                                     d: 2064855961)
+
+      assert_instance_of OpenSSL::PKey::RSA, new_key
+      assert_equal true, new_key.private?
+      assert_equal OpenSSL::BN.new(3161751493), new_key.n
+      assert_equal OpenSSL::BN.new(65537), new_key.e
+      assert_equal OpenSSL::BN.new(2064855961), new_key.d
+    end
+
+    def test_s_from_parameters_rsa_with_n_e_and_d_given
+      new_key = OpenSSL::PKey.from_parameters("RSA", "n" => OpenSSL::BN.new(3161751493),
+                                                     "e" => OpenSSL::BN.new(65537),
+                                                     "d" =>  OpenSSL::BN.new(2064855961))
+
+      assert_instance_of OpenSSL::PKey::RSA, new_key
+      assert_equal true, new_key.private?
+      assert_equal OpenSSL::BN.new(3161751493), new_key.n
+      assert_equal OpenSSL::BN.new(65537), new_key.e
+      assert_equal OpenSSL::BN.new(2064855961), new_key.d
+    end
+
+    def test_s_from_parameters_rsa_with_n_and_e_given
+      new_key = OpenSSL::PKey.from_parameters("RSA", n: OpenSSL::BN.new(3161751493),
+                                                     e: OpenSSL::BN.new(65537))
+
+      assert_instance_of OpenSSL::PKey::RSA, new_key
+      assert_equal false, new_key.private?
+      assert_equal OpenSSL::BN.new(3161751493), new_key.n
+      assert_equal OpenSSL::BN.new(65537), new_key.e
+      assert_equal nil, new_key.d
+    end
+
+    def test_s_from_parameters_rsa_with_openssl_internal_names
+      source  = OpenSSL::PKey::RSA.generate(2048)
+      new_key =  OpenSSL::PKey.from_parameters("RSA", n: source.n,
+                                                      e: source.e,
+                                                      d: source.d,
+                                                      "rsa-factor1" => source.p,
+                                                      "rsa-factor2" => source.q,
+                                                      "rsa-exponent1" => source.dmp1,
+                                                      "rsa-exponent2" => source.dmq1,
+                                                      "rsa-coefficient1" => source.iqmp
+                                                    )
```

Comment:
The thing for the `)` is same with the one I mentioned in another part.

Reply:
Adjusted the alignment of everything